src/dataflow/src/source/mod.rs

```diff
@@ -1361,43 +1395,56 @@ where
                     //
                     // Side note: with persistence enabled for the source, we will anyways never
                     // re-emit old data, because we start reading from the persisted offsets.
-                    let current_binding = timestamp_histories
-                        .get_binding(&source_ts.partition, source_ts.offset)
-                        .map(|(binding, _offset)| binding)
-                        .unwrap_or(0);
-                    if current_binding < assigned_ts.0 {
-                        timestamp_histories.add_binding(
-                            source_ts.partition.clone(),
-                            assigned_ts.0,
-                            source_ts.offset,
-                            false,
-                        );
-                    } else {
-                        tracing::debug!(
-                            "Filtered out timestamp binding {:?} from persistence because we already have {}.",
-                            (source_ts, assigned_ts),
-                            current_binding
-                        );
+                    let lower = Antichain::from_elem(assigned_ts.0);
+                    let upper = Antichain::new();
+                    // NOTE: We use get_bindings_in_range() and not the seemingly better
+                    // get_binding(). However, get_binding() does not only consult the actual
+                    // stored bindings but also checks the Proposer, and returns its output as a
+                    // binding if it doesn't have a stored binding. We only care about stored
+                    // bindings here.
+                    let existing_binding = timestamp_histories
+                        .get_bindings_in_range(lower.borrow(), upper.borrow()).into_iter()
+                        .filter(|(pid, _ts, _offset)| *pid == source_ts.partition).next();
+                    match existing_binding {
+                        None => {
+                            timestamp_histories.add_binding(
+                                source_ts.partition.clone(),
+                                assigned_ts.0,
+                                source_ts.offset,
+                                false,
+                            );
+                        },
+                        Some(existing_binding) => {
```

Comment:
if the existing binding is not identical to the one in `persist` that will mess up the retraction/insertion logic right?

if say instead of `(a, b)` we get `(a, c)` from the sqlite bindings then we will end up retracting `(a, b)` and inserting `(a, c)`. Maybe that's fine I just wanted to call it out and see if was indeed fine?

Reply:
It is fine, yes. The retraction logic will see that the current contents of `timestamp_histories` does not match what it thinks is stored in persistence, and will retract updates to bring persist in line with the current bindings.

I think. :crossed_fingers: (it will be much easier once we don't allow mixing persist and SQLite...)